File: vi/codes/ruby/chapter_dynamic_programming/climbing_stairs_dfs.rb
=begin
File: climbing_stairs_dfs.rb
Created Time: 2024-05-29
Author: Xuan Khoa Tu Nguyen (ngxktuzkai2000@gmail.com)
=end

### Tìm kiếm ###
def dfs(i)
  # Đã biết dp[1] và dp[2], trả về chúng
  return i if i == 1 || i == 2
  # dp[i] = dp[i-1] + dp[i-2]
  dfs(i - 1) + dfs(i - 2)
end

### Leo cầu thang: tìm kiếm ###
def climbing_stairs_dfs(n)
  dfs(n)
end

### Mã chương trình chính ###
if __FILE__ == $0
  n = 9

  res = climbing_stairs_dfs(n)
  puts "Climbing #{n} stairs has #{res} solutions"
end

File: vi/codes/csharp/chapter_dynamic_programming/climbing_stairs_dfs.cs
/**
* File: climbing_stairs_dfs.cs
* Created Time: 2023-06-30
* Author: hpstory (hpstory1024@163.com)
*/

namespace hello_algo.chapter_dynamic_programming;

public class climbing_stairs_dfs {
    /* Tìm kiếm */
    int DFS(int i) {
        // Đã biết dp[1] và dp[2], trả về chúng
        if (i == 1 || i == 2)
            return i;
        // dp[i] = dp[i-1] + dp[i-2]
        int count = DFS(i - 1) + DFS(i - 2);
        return count;
    }

    /* Leo cầu thang: Tìm kiếm */
    int ClimbingStairsDFS(int n) {
        return DFS(n);
    }

    [Test]
    public void Test() {
        int n = 9;
        int res = ClimbingStairsDFS(n);
        Console.WriteLine($"Leo {n} bậc thang có {res} lời giải");
    }
}
